com.unity.ml-agents/Runtime/Sensor/RayPerceptionSensor.cs

```diff
@@ -4,91 +4,245 @@
             /// </summary>
-            public int age
+            public bool hasHit;
+
+            /// <summary>
+            /// Whether or not the ray hit an object whose tag is in the input's detectableObjects list.
+            /// </summary>
+            public bool hitTaggedObject;
+
+            /// <summary>
+            /// The index of the hit object's tag in the detectableObjects list, or -1 if there was no hit, or the
+            /// hit object has a different tag.
+            /// </summary>
+            public int hitTagIndex;
+
+            /// <summary>
+            /// Normalized distance to the hit object.
+            /// </summary>
+            public float hitFraction;
+        }
+
+        /// <summary>
+        /// Writes the ray output information to a float array.  Each element in the rayAngles array determines a
+        /// sublist of data to the observation. The sublist contains the observation data for a single cast.
+        /// The list is composed of the following:
+        /// 1. A one-hot encoding for detectable objects. For example, if detectableObjects.Length = n, the
+        ///    first n elements of the sublist will be a one-hot encoding of the detectableObject that was hit, or
+        ///    all zeroes otherwise.
+        /// 2. The 'numDetectableObjects' element of the sublist will be 1 if the ray missed everything, or 0 if it hit
+        ///    something (detectable or not).
+        /// 3. The 'numDetectableObjects+1' element of the sublist will contain the normalized distance to the object
+        ///    hit, or 1.0 if nothing was hit.
+        /// </summary>
+        /// <param name="numDetectableObjects"></param>
+        /// <param name="buffer">Output buffer. The size must be equal to (numDetectableObjects+2) * rayOutputs.Length</param>
+        public void ToFloatArray(int numDetectableObjects, float[] buffer)
+        {
+            Array.Clear(buffer, 0, buffer.Length);
+            var bufferOffset = 0;
+
+            for (var i = 0; i < rayOutputs.Length; i++)
             {
-                get { return Time.frameCount - m_Frame; }
+                var rayOutput = rayOutputs[i];
+                if (rayOutput.hitTaggedObject)
+                {
+                    buffer[bufferOffset + rayOutput.hitTagIndex] = 1f;
+                }
+                buffer[bufferOffset + numDetectableObjects] = rayOutput.hasHit ? 0f : 1f;
+                buffer[bufferOffset + numDetectableObjects + 1] = rayOutput.hitFraction;
+                bufferOffset += numDetectableObjects + 2;
             }
+        }
 
-            public RayInfo[] rayInfos;
+        /// <summary>
+        /// RayOutput for each ray that was cast.
+        /// </summary>
+        public RayOutput[] rayOutputs;
+    }
+
+    /// <summary>
+    /// Debug information for the raycast hits. This is used by the RayPerceptionSensorComponent.
+    /// </summary>
+    internal class DebugDisplayInfo
```

Comment:
Moved out of RayPerceptionSensor, now internal instead of public.

Reply:
I was debating merging this with RayPerceptionOutput, but it would bloat that with fields that we don't always need (start, end, and castRadius) and extra logic for the age (which I just use for decaying the alpha during display).

Reply:
It makes sense to keep it separated, as the Offspring sang about.  

Seems like `RayInfo` is a view of `RayPerceptionInput`.  Do you see any commonalities there?